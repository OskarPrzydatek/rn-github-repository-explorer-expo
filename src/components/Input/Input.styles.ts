import { colors } from '@/constants';
import { StyleSheet } from 'react-native';

export const inputStyles = StyleSheet.create({
  input: {
    height: 40,
    padding: 12,
    borderRadius: 6,
  },
  blurBorder: {
    borderWidth: 1,
    borderColor: colors.darkGrey,
    backgroundColor: colors.grey,
  },
  focusBorder: {
    borderWidth: 2,
    borderColor: colors.blue,
  },
});
