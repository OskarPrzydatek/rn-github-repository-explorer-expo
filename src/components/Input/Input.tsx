import React, { useState } from 'react';

import { Text, TextInput, View } from 'react-native';

import { inputStyles } from '@/components/Input/Input.styles';
import { colors } from '@/constants';

interface IInput {
  value: string;
  placeholder: string;
  isValidationError: boolean;
  isValidationErrorMessage: string;
  onChangeText: (text: string) => void;
  onSubmitEditing: () => void;

  testID?: string;
  validationMessageTestID?: string;
}

export const Input = ({
  value,
  placeholder,
  isValidationError,
  isValidationErrorMessage,
  onChangeText,
  onSubmitEditing,

  testID,
  validationMessageTestID,
}: IInput) => {
  const [isFocused, setIsFocused] = useState(false);

  const isFocusedStyle = isFocused
    ? inputStyles.focusBorder
    : inputStyles.blurBorder;

  const onFocus = () => setIsFocused(true);
  const onBlur = () => setIsFocused(false);

  return (
    <View style={inputStyles.input}>
      <TextInput
        value={value}
        style={[inputStyles.textInput, isFocusedStyle]}
        selectionColor={colors.black}
        placeholder={placeholder}
        onFocus={onFocus}
        onBlur={onBlur}
        onChangeText={onChangeText}
        onSubmitEditing={onSubmitEditing}
        testID={testID}
      />

      {isValidationError && (
        <Text style={inputStyles.errorMessage} testID={validationMessageTestID}>
          {isValidationErrorMessage}
        </Text>
      )}
    </View>
  );
};
