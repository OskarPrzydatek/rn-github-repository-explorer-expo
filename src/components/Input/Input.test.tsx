import { fireEvent, render, screen } from '@testing-library/react-native';

import { Input } from '@/components/Input/Input';
import { inputStyles } from '@/components/Input/Input.styles';

describe('Input', () => {
  it('should match snapshot', () => {
    const tree = render(
      <Input
        value={''}
        placeholder={''}
        isValidationError={false}
        isValidationErrorMessage={''}
        onChangeText={jest.fn()}
        onSubmitEditing={jest.fn()}
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('should render validation message when is validation error', () => {
    render(
      <Input
        value={''}
        placeholder={''}
        isValidationError={true}
        isValidationErrorMessage="Validation error message"
        onChangeText={jest.fn()}
        onSubmitEditing={jest.fn()}
        validationMessageTestID="input-validation-message"
      />,
    );

    const validationErrorMessage = screen.getByTestId(
      'input-validation-message',
    );

    expect(validationErrorMessage).not.toBeNull();
  });

  it('should change styles when focus', () => {
    render(
      <Input
        value={''}
        placeholder={''}
        isValidationError={true}
        isValidationErrorMessage="Validation error message"
        onChangeText={jest.fn()}
        onSubmitEditing={jest.fn()}
        testID="input-component"
      />,
    );

    const input = screen.getByTestId('input-component');

    fireEvent(input, 'focus');

    expect(input.props.style).toEqual([
      inputStyles.textInput,
      inputStyles.focusBorder,
    ]);

    fireEvent(input, 'blur');

    expect(input.props.style).toEqual([
      inputStyles.textInput,
      inputStyles.blurBorder,
    ]);
  });
});
