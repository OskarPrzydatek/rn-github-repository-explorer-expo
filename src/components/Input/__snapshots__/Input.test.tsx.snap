// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot 1`] = `
<View
  style={
    {
      "gap": 8,
    }
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[MockFunction]}
    onFocus={[Function]}
    onSubmitEditing={[MockFunction]}
    placeholder=""
    selectionColor="#1f2329"
    style={
      [
        {
          "borderRadius": 6,
          "height": 40,
          "padding": 12,
        },
        {
          "backgroundColor": "#f6f8fa",
          "borderColor": "#d0d7de",
          "borderWidth": 1,
        },
      ]
    }
    value=""
  />
</View>
`;
